bazel-build.el

```diff
@@ -0,0 +1,196 @@
+         (string-prefix (car string-split))
+         (string-base (cadr string-split))
+         (filename (file-name-nondirectory string-base))
+         (dirname (or (file-name-directory string-base) ""))
+         (slashpos (or (string-match "/.*$" string-base) 0))
+		 (workspace-root (bazel-build--find-workspace-root dirname))
+         (subdirectories (bazel-build--subdirectory-completions dirname workspace-root))
+         (rules (bazel-build--rule-completions dirname workspace-root))
+         (candidate-completions (mapcar (lambda (d) (concat string-prefix d)) (append subdirectories rules)))
+         (completions (all-completions string candidate-completions)))
+
+      ;; return differently based on what mode we were called in
+      ;; see http://www.gnu.org/software/emacs/manual/html_node/elisp/Programmed-Completion.html
+      (cond
+       ;; `try-completion' operation.
+       ;;  - if the specified string is a unique and exact match, return t
+       ;;  - if there is more than one match, return the common substring of all
+       ;;    matches. if the string is an exact match for one completion
+       ;;    alternative but also matches other longer alternatives, the return
+       ;;    value is the string.
+       ;;  - if there are no matches, it should return nil.
+       ((eq mode nil)
+        (cond
+         ((eq 0 (length completions)) nil)
+
+         ((and (eq 1 (length completions)) (string= string (car completions))) 't)
+
+         ('t (reduce #'fill-common-string-prefix completions))))
+
+       ;; `all-completions' operation. returns a list of all possible completions.
+       ((eq mode 't) completions)
+
+       ;; `test-completion' operation.
+       ;;  - if the specified string is an exact match for some completion
+       ;;    alternative, returns t.
+       ;;  - otherwise, returns nil.
+       ((eq mode 'lambda) (member (file-name-as-directory filename) completions))
+
+       ;; `completion-boundaries' operation.
+       ;; let completion know our completions only apply after the last /
+       ('t (cons (list 'boundaries slashpos) (length filename))))))
+
+(defun bazel-build--suggest-initial-input (filename workspace-root)
+  "Suggest initial input to the interactive completion.
+FILENAME is the filename of the current buffer.
+WORKSPACE-ROOT is the root of the Bazel workspace."
+  (concat "//" (directory-file-name (bazel-build--extract-build-path filename workspace-root))))
+
+(defun bazel-build ()
+  "Find an run a Bazel build rule."
+  (interactive)
+  (let* ((filename (buffer-file-name))
+         (workspace-root (bazel-build--find-workspace-root filename))
+         (initial-input (bazel-build--suggest-initial-input filename workspace-root))
+		 (target (completing-read
+				"bazel build "             ; prompt
+				#'bazel-build--completions ; collection
+				nil                        ; predicate
+				't                         ; require-match
+				initial-input)))           ; initial-input
```

Comment:
I'm afraid that achieving the same completion behavior using default instead of initial-input would require a fairly substantial change to this code. What do you think about saving it for a follow-up PR?

Reply:
Ah, saw this just now. A follow-up sounds good. Consider adding a TODO here.